flytekit/core/promise.py

```diff
@@ -1066,32 +1078,22 @@ def create_and_link_node(
 
     for k in sorted(interface.inputs):
         var = typed_interface.inputs[k]
+        if var.type.simple == SimpleType.NONE:
+            raise TypeError("Arguments do not have type annotation")
         if k not in kwargs:
-            is_optional = False
-            if var.type.union_type:
-                for variant in var.type.union_type.variants:
-                    if variant.simple == SimpleType.NONE:
-                        val, _default = interface.inputs_with_defaults[k]
-                        if _default is not None:
-                            raise ValueError(
-                                f"The default value for the optional type must be None, but got {_default}"
-                            )
-                        is_optional = True
-            if not is_optional:
-                from flytekit.core.base_task import Task
-
+            # interface.inputs_with_defaults[k][0] is the type of the default argument
+            # interface.inputs_with_defaults[k][1] is the value of the default argument
+            if k in interface.inputs_with_defaults and (
+                interface.inputs_with_defaults[k][1] is not None
+                or UnionTransformer.is_optional_type(interface.inputs_with_defaults[k][0])
+            ):
+                default_val = interface.inputs_with_defaults[k][1]
+                if not isinstance(default_val, Hashable):
+                    raise _user_exceptions.FlyteAssertion("Cannot use non-hashable object as default argument")
```

Comment:
Why do we need this extra-check? Any value that can't be handled by the existing type transformers at registration time is going to show pretty prominently, right?

Reply:
Relying on `Hashable` prevents dataclasses from being used as default arguments. This is too restrictive, since even sophisticated linters like ruff are not able to catch this example:
```
from dataclasses import dataclass
from mashumaro import DataClassDictMixin

@dataclass
class DC(DataClassDictMixin):
    a: int
    b: str

def f(dc: DC = DC(42, "43"), xs: list[int] = [1,2], d: dict[str, int] = {"abc": 42}, s: set(int) = set([1,2,3])):
    ...

if __name__ == "__main__":
    f()

```
Running ruff with https://docs.astral.sh/ruff/rules/mutable-argument-default/ turned on returns:
```
❯ ruff check example
example/dc.py:9:46: B006 Do not use mutable data structures for argument defaults
   |
 7 |     b: str
 8 |
 9 | def f(dc: DC = DC(42, "43"), xs: list[int] = [1,2], d: dict[str, int] = {"abc": 42}, s: set(int) = set([1,2,3])):
   |                                              ^^^^^ B006
10 |     ...
   |
   = help: Replace with `None`; initialize within function

example/dc.py:9:73: B006 Do not use mutable data structures for argument defaults
   |
 7 |     b: str
 8 |
 9 | def f(dc: DC = DC(42, "43"), xs: list[int] = [1,2], d: dict[str, int] = {"abc": 42}, s: set(int) = set([1,2,3])):
   |                                                                         ^^^^^^^^^^^ B006
10 |     ...
   |
   = help: Replace with `None`; initialize within function

example/dc.py:9:100: B006 Do not use mutable data structures for argument defaults
   |
 7 |     b: str
 8 |
 9 | def f(dc: DC = DC(42, "43"), xs: list[int] = [1,2], d: dict[str, int] = {"abc": 42}, s: set(int) = set([1,2,3])):
   |                                                                                                    ^^^^^^^^^^^^ B006
10 |     ...
   |
   = help: Replace with `None`; initialize within function

Found 3 errors.
No fixes available (3 hidden fixes can be enabled with the `--unsafe-fixes` option).
```

I propose we handle these 3 common errors explicitly instead. This is also going to simplify the error message. I'll get a PR out soon.

Reply:
FYI: https://github.com/flyteorg/flytekit/pull/2651

Reply:
@eapolinario I saw your PR. Quite make senses for me. Those 3 container types are indeed the most commonly used non-hashable types.

Reply:
How about `collections.deque`?